### DISCUSSION TITLE
STM32F205xx and SDIO support

### DISCUSSION THREAD

**Discussion by xmacur06** [zephyrproject-rtos/zephyr#68877] [Q&A] [answered]
Hello,
Zephyr version: 3.5.99
zephyr-sdk-0.16.4

I have custom board with stm32f205 and connected sd card via SDIO interface. 
Because of missing DT node sdio in stm32f2.dtsi I added this node into my board device tree like this:

```
soc {
        sdmmc1: sdmmc@40012c00 {
            compatible = "st,stm32-sdmmc";
            reg = < 0x40012C00 0x400 >;
            clocks = < &rcc STM32_CLOCK_BUS_APB2 0x00000800 >;
            resets = < &rctl STM32_RESET(APB2, 11U) >;
            interrupts = < 49 0 >;
            status = "okay";
            pinctrl-0 = < &sdio_default >;
            pinctrl-names = "default";

            mmc {
                    compatible = "zephyr,sdmmc-disk";
                    status = "okay"; 
            };
        };
 };
```

My prj.conf looks like:
```
CONFIG_DISK_DRIVERS=y
CONFIG_DISK_ACCESS=y
CONFIG_DISK_DRIVER_SDMMC=y
CONFIG_SDMMC_STM32=y
CONFIG_FILE_SYSTEM=y
```
But build is failing with this error and warning.

> .
> .
> .
> CMake Warning at /home/jm/zephyrproject/zephyr/CMakeLists.txt:853 (message):
>   No SOURCES given to Zephyr library: drivers__sdhc
> 
>   Excluding target from build.
> .
> .
> .
> FAILED: zephyr/zephyr_pre0.elf zephyr/zephyr_pre0.map /home/jm/apps/caline_cp33/build/zephyr/zephyr_pre0.map 
> : && ccache /home/jm/zephyr-sdk-0.16.4/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc  -gdwarf-4 zephyr/CMakeFiles/zephyr_pre0.dir/misc/empty_file.c.obj -o zephyr/zephyr_pre0.elf  zephyr/CMakeFiles/offsets.dir/./arch/arm/core/offsets/offsets.c.obj  -fuse-ld=bfd  -T  zephyr/linker_zephyr_pre0.cmd  -Wl,-Map=/home/jm/apps/caline_cp33/build/zephyr/zephyr_pre0.map  -Wl,--whole-archive  app/libapp.a  zephyr/libzephyr.a  zephyr/arch/common/libarch__common.a  zephyr/arch/arch/arm/core/libarch__arm__core.a  zephyr/arch/arch/arm/core/cortex_m/libarch__arm__core__cortex_m.a  zephyr/lib/libc/picolibc/liblib__libc__picolibc.a  zephyr/lib/libc/common/liblib__libc__common.a  zephyr/subsys/fs/libsubsys__fs.a  zephyr/subsys/sd/libsubsys__sd.a  zephyr/drivers/disk/libdrivers__disk.a  zephyr/drivers/interrupt_controller/libdrivers__interrupt_controller.a  zephyr/drivers/clock_control/libdrivers__clock_control.a  zephyr/drivers/gpio/libdrivers__gpio.a  zephyr/drivers/pinctrl/libdrivers__pinctrl.a  zephyr/drivers/reset/libdrivers__reset.a  zephyr/drivers/timer/libdrivers__timer.a  modules/stm32/stm32cube/lib..__modules__hal__stm32__stm32cube.a  -Wl,--no-whole-archive  zephyr/kernel/libkernel.a  -L"/home/jm/zephyr-sdk-0.16.4/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.2.0/thumb/v7-m/nofp"  -L/home/jm/apps/caline_cp33/build/zephyr  -lgcc  zephyr/arch/common/libisr_tables.a  -mcpu=cortex-m3  -mthumb  -mabi=aapcs  -mfp16-format=ieee  -mtp=soft  -Wl,--gc-sections  -Wl,--build-id=none  -Wl,--sort-common=descending  -Wl,--sort-section=alignment  -Wl,-u,_OffsetAbsSyms  -Wl,-u,_ConfigAbsSyms  -nostdlib  -static  -Wl,-X  -Wl,-N  -Wl,--orphan-handling=warn  -Wl,-no-pie  -DPICOLIBC_LONG_LONG_PRINTF_SCANF  --specs=picolibc.specs  -lc  -lgcc && cd /home/jm/apps/caline_cp33/build/zephyr && /usr/bin/cmake -E true
> /home/jm/zephyr-sdk-0.16.4/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.2.0/../../../../arm-zephyr-eabi/bin/ld.bfd: modules/stm32/stm32cube/lib..__modules__hal__stm32__stm32cube.a(stm32f2xx_hal_sd.c.obj): in function `HAL_SD_IRQHandler':
> /home/jm/zephyrproject/modules/hal/stm32/stm32cube/stm32f2xx/drivers/src/stm32f2xx_hal_sd.c:1642: undefined reference to `HAL_DMA_Abort_IT'
> /home/jm/zephyr-sdk-0.16.4/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.2.0/../../../../arm-zephyr-eabi/bin/ld.bfd: /home/jm/zephyrproject/modules/hal/stm32/stm32cube/stm32f2xx/drivers/src/stm32f2xx_hal_sd.c:1632: undefined reference to `HAL_DMA_Abort_IT'
> collect2: error: ld returned 1 exit status

**Comment by FRASTM** [comment]
hi @xmacur06 
maybe declaration of weak function in the zephyr driver can help
```
 /* weak function required for HAL compilation */
__weak HAL_StatusTypeDef HAL_DMA_Abort_IT(DMA_HandleTypeDef *hdma)
{
	return HAL_OK;
}

/* weak function required for HAL compilation */
__weak HAL_StatusTypeDef HAL_DMA_Abort(DMA_HandleTypeDef *hdma)
{
	return HAL_OK;
}
```

**Comment by FRASTM** [comment]
note that DMA2 of the stm32f205 has a channel/request for SDIO transfer if needed

**Comment by xmacur06** [comment]
Hello and thank you for responce. 
I found the issue why build failed (main issue DMA_LL driver used instead of DMA_HAL driver). Problem (and maybe bug) is that kconfig `CONFIG_SDMMC_STM32` depends on `DT_HAS_ST_STM32_SDMMC_ENABLED && CONFIG_DISK_DRIVER_SDMMC && CONFIG_DISK_DRIVERS` it is fine. But at this point `CONFIG_SDMMC_STM32` also selects `CONFIG_USE_STM32_HAL_DMA if CONFIG_SOC_SERIES_STM32L4X || CONFIG_SOC_SERIES_STM32F7X || CONFIG_SOC_SERIES_STM32F4X` and I don't see STM32F2X series.

So in my Kconfig.defconfig I added: 

```
config MYAPP_STM32
  default y
  bool
  select USE_STM32_HAL_DMA if DT_HAS_ST_STM32_SDMMC_ENABLED
```

So I think that  `CONFIG_USE_STM32_HAL_DMA if CONFIG_SOC_SERIES_STM32L4X || CONFIG_SOC_SERIES_STM32F7X || CONFIG_SOC_SERIES_STM32F4X` should be extend by F2 series.

And also default device tree st/f2/stm32f2.dtsi should be extened by sdmmc/sdio node.

**Comment by FRASTM** [comment]
I confirm 
That could still be an enhancement for the stm32F2 serie,
( but some of the changes are about your custom board)

**Comment by xmacur06** [comment]
Maybe there is some misconception. I think this is general problem for STM32F2X series with usage of SDIO driver. SDIO always uses HAL driver which calls HAL_DMA functions. But without fix, LL_DMA is build instead of HAL_DMA and linker script is never successful.

So it is not bug just for my custom board but for all of them with F2 + SDIO.

**Reply by erwango** [reply, marked as answer]
So it is not a bug but a feature not yet supported on STM32F2 series, you can raise an issue for this.
Alternatively you can also work this out and submit a contribution, which would be welcome.

**Comment by xmacur06** [comment]
Thank you for response Erwan. 
You are right, it is not bug but unsupported feature. 
And yes I'll try contribute after studying of Contribution Guidelines.

**Reply by prayassamriya** [reply]
Hi @xmacur06, Are you actively working on adding SDIO support for STM?

**Reply by xmacur06** [reply]
Hello, yes currently I do some investigation. 
Not all is still clear for me in sense of DT binding + CMAKE + Kconfig.
